### DISCUSSION TITLE
Credit Card Handling

### DISCUSSION THREAD

**Discussion by Medicliffy** [actualbudget/releases#28] [Feature ideas]
I'll start by saying that I'm new to the community (another budding convert from a similar program...) and have only explored a little bit with the demo, but I'm really excited about the approach and priorities I've heard expressed. Looking to solve use cases rather than copy features, taking the needed time to build good bones even though it delays short term release of features, local first, having a simple front-end but also really powerful tools like rules, etc.

I'm essentially posting because I think I see the credit card use case that's missing and how it compares to how it is handled in nYNAB. Hopefully this actually makes sense and is helpful in considering how credit cards might be handled in Actual.

### Use Cases
I think the use cases boil down to:
1. "Assign" money to paying the credit card from currently budgeted spending (e.g. I pay $20 for gas on my card, my gas budget goes down by $20, and I have an additional $20 that's earmarked for paying the credit card)
2. Pay down an existing balance on the credit card (for those carrying debt on the card)
3. Differentiate between the above
4. Update the balance w/ interest

### My understanding of actual
If I understand how Actual manages credit cards, I think it essentially can improve in how it handles use cases 2 & 3 (as stated in the live stream that carrying debt is where the management could be improved).

For those not carrying debt, the credit card effectively starts at $0, new transactions count as a negative balance on the card and impact the category (such as gas) as they should, and the amount available to pay the card is exactly what the balance on the card is as shown in the account screen. So, say you budged $30 to gas then spend $20 on a card. Now gas has a $10 balance, the card is -$20, and the $30 of money that you budgeted is still in whatever accounts it started in. Now you pay off the card, the cash is transferred making the card $0 and the cash $10, which matches the balance in the category. Thus everything adds up at all points in time and all is well. There's nothing really to "assign" for 1, and 2 & 3 don't apply.

For those carrying credit card debt it's tricky. The livestream showed creating a credit card debt category (would need to be for each card with debt), assigning the debt to that (with the starting balance transaction assigned to CC A Debt - so a $250 starting balance on the first month would have a CC A debt category with $0 budgeted, -$250 spent, and -$250 balance), and carrying the balance over, and that works. Here are where I think the challenges are:
1. This works just like above. See 3 about differentiating.
2. This works, but I think it's different than shown in the livestream. I don't think any workflow would actually have income directly to the credit card (except in the case of credit card rewards, I suppose, but not for paying off debt), which is what was shown. However, what you're effectively doing is budgeting some money you have to pay off the credit card debt - so do just that, add money to "budged" for the credit card. This appropriately reduces the debt you're carrying - you have to make sure you transfer that amount, though. So this requires 2 separate actions which may not seem as related as one would like.
3. Here is the biggest challenge, I think. To pay off current spending on a credit card (which should be immediate if using the budget to guide spending) it's not clear how much money that is. You can find it by taking the current credit card account balance and subtracting the CC A Debt category balance. With the above example, let's say we start with $250 in CC debt on card A, budget $30 of money to gas, and spend $20 on gas. Our CC A debt category remains -250 while CC A account becomes -270. The difference ($20) is the amount of current spending that we budgeted for (in this case on gas), so it's important to pay off that $20 so you don't fall further behind - and, besides that, it's what you already budgeted to spend.
4. This is handled essentially the same way as the starting balance - A transaction on the credit card with payee "interest" in the category "CC A Debt" with a "payment" of the amount of interest (in which case "payment" is confusing since a "credit card payment" is actually a deposit into the credit card based on the column names in Actual). To pay for the interest and avoid falling further behind, it's the same process as number 2. For, say, $5 of interest, transfer $5 from your bank account to your credit card (in real life paying the interest and also in Actual reflecting that transfer), and ALSO budget $5 into the CC A Debt category to reflect that $5 is no longer available to spend elsewhere.
Note that this process is also needed if you start Actual with a balance on your credit card. Say you bought that $20 of gas before starting Actual. You'd then need to start with $20 of CC A Debt, and budget your $30 with $10 to gas and $20 to CC A Debt (since Actual doesn't know where your $20 balance on CC A came from it has no connection to the gas at all, you just know that you now only need $10 of gas for the rest of the month and you were planning on paying off that $20).

### How nYNAB handled/handles it
At least prior to a recent revamping of debt management (I'm not sure if or how that changed this process). nYNAB automatically made credit card categories for each credit card account you made. This category is basically built to show you how much you should pay on your credit card based on your current budget. So the matching process to what I described with Actual above is to - add your credit card with the initial $250 balance and the YNAB behavior is to:
- Show your CC A account with -$250
- Make a CC A category in a "Credit Card Payments" category group
- Show (in Actual terminology) 0 budgeted, 0 spent, and 0 balance - in YNAB instead of "balance" they show "available"

Then you use the credit card for $20 on gas. YNAB:
- Shows $20 spent in gas, just like Actual
- Adds an equal/opposite transaction showing "spent" money in the CC A category - this makes the "available" amount $20, which makes it clear that this amount is what you should cover based on your current budget
- You have to add money into "budgeted" (again, Actual terminology as opposed to YNAB) to pay down the interest/balance
- The revolving CC balance you have ($250, in this example) is not shown anywhere (like it is in Actual when you add the "CC A Debt" category).

This process in YNAB works the same for not carrying a CC balance. You just never have to budget any to the "CC A" category (except for the balance you started with, like if you started after spending that $20 on gas) because the behavior when you spend budgeted money using the credit card updates the amount "available" to make a CC payment for you.
Then, when you transfer money from your checking account (for example) to CC A, 

### Another consideration: CC rewards
CC rewards are really confusing in nYNAB. You add it as an inflow to the "CC A" category and it makes more available for the payment than the balance on your card (because the reward doesn't actually show up as money for you to budget...because it handles credit cards in such a weird way). You then have to budget a negative amount into the CC A category to get access to that money to budget. I know that sounds sketchy (it's the main thing I really dislike about how they handle CCs), but it makes sense when you realize that you earmarked $20 for gas that exists in your checking account, then you only had to spend $15 of that earmarked $20 because of your $5 reward. So you legitimately have that extra $5 to budget.

In contrast, it's super straightforward in Actual. Just put the CC reward as income on the CC, it shows up income in whatever category you put it in, and it increases your "to budget" amount, and you're done.

### My thoughts on a solution
In short, don't make credit cards first class citizens, but do make "debt accounts" first class citizens (at least I think that would be needed for my suggested solution). I think all of the above use cases apply to other debts, credit cards are just somewhat unique in that they're a revolving door and don't necessarily represent revolving debt. This would require some type of linking an account (the CC, mortgage, HELOC, whatever) to a category in the debt group (which should be made automatically when you make a new debt category).

When you want to decrease the debt or pay interest, you should budget directly into that debt category. When you're making purchases that are budgeted for but temporarily paid with that debt (such as using a credit card for gas), that should somehow be tracked so that you know how much money you should pay on your debt/CC to stay in the same place on the treadmill. Essentially, you should have somewhere to look to see $20 in my example above, rather than having to look at the difference between your CC A debt category and you CC A account balance.

I'm not sure the best way to do that, but I think that gets to the heart of the problem in the most generalized way. 

### My actual idea
You should be able to see: your revolving debt (if there is any), your current balance, and your current amount set aside for payment (both from current transactions and your planned amount paying both principle and interest). Maybe the simplest way would be making a third group: "Debt" that's separated like "Income" is. It would also be unique in that each category would have to have a corresponding account (might have to be on budget accounts, haven't thought that far though). It would have columns for budgeted, interest (and starting principle), revolving balance, and "available to pay". The first 3 would behave as they currently do (budgeted = budgeted, interest (and starting principle) = spent, revolving balance = balance, and "available to pay" = revolving balance - balance of the connected account). Obviously "interest (and starting principle)" is a terrible name (and the others might not be the best either - revolving balance might better stated capturing that it's "debt that's rolling over to next month"). Also, there'd likely be a more reliable way to calculate "available to pay" (might even consider if it should alert when you've overspent another category with a credit card...though that's probably already handled sufficiently as is). CC rewards would not show up in that category except by essentially moving some "available to pay" back to "to budget", and I'm not sure what the optimal view would be if you clicked on "available to pay" (ideally it would somehow summarize the components that went into it, which would be the amount budgeted that month, the amount budgeted/spent on other categories with that card that month, and any rewards/other inflows into the account).

Sheesh, that ended up being far longer than I expected. Hopefully it's reasonably intelligible and helpful in planning how to make the most clear UI/UX for credit cards or for those working on how to manage them as currently implemented in Acutal!

**Comment by bpaulien** [comment]
Yes, that seems to capture the essence of the problem.  I'm "on the CC float" myself, so have some $s accounted for in the debt category, and some not, and I'm in the process of trying to work out the right workflow in Actual to accurately reflect a "pay statement balance in full" payment, so both transfer and possibly category update happen accurately.  Right now it seems like some manual calculations would be needed to figure how much to move into/out of the associated debt category.
